.github/workflows/compatability-tests.yaml

```diff
@@ -0,0 +1,300 @@
+
+      - name: Generate code
+        run: |
+          make -j3 gen-api VERSION=${{ steps.version.outputs.tag }}
+          mkdir webui/dist
+          touch webui/dist/index.html
+          tar -cf /tmp/generated.tar.gz .
+
+      - name: Store generated code
+        uses: actions/upload-artifact@v3
+        with:
+            name: generated-code
+            path: /tmp/generated.tar.gz
+
+#      - name: Package Spark App
+#        working-directory: test/spark/app
+#        run: sbt sonnets-311/package
+#
+#      - name: Store Sonnets Spark app
+#        uses: actions/upload-artifact@v3
+#        with:
+#          name: generated-sonnets
+#          path: test/spark/app/target/sonnets-311/target/sonnets-311/scala-2.12/sonnets-311_2.12-0.1.0.jar
+
+      - name: Build Spark direct-access client
+        working-directory: clients/hadoopfs
+        run: |
+          mvn -Passembly -Djar.finalName=client -DskipTests --batch-mode --update-snapshots package
+
+      - name: Store client assembly
+        uses: actions/upload-artifact@v3
+        with:
+          name: client-assembly
+          path: clients/hadoopfs/target/client.jar
+
+  deploy-image:
+    name: Build and cache Docker image
+    needs: [gen-code]
+    runs-on: ubuntu-20.04
+    outputs:
+      tag: ${{ steps.version.outputs.tag }}
+      image_id: ${{ steps.build_export.outputs.ImageID }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Setup NodeJS
+        uses: actions/setup-node@v3
+        with:
+          node-version: "16.17.1"
+
+      - name: Setup Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: 1.19.2
+        id: go
+
+      - name: Retrieve generated code
+        uses: actions/download-artifact@v3
+        with:
+          name: generated-code
+          path: /tmp/
+
+      - name: Unpack generated code
+        run: tar -xf /tmp/generated.tar.gz
+
+      - name: Extract version
+        shell: bash
+        run: echo "tag=sha-$(git rev-parse --short HEAD | sed s/^v//g)" >> $GITHUB_OUTPUT
+        id: version
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Build and export
+        uses: docker/build-push-action@v2
+        id: build_export
+        with:
+          context: .
+          file: ./Dockerfile
+          tags: treeverse/lakefs:${{ steps.version.outputs.tag }}
+          outputs: type=docker,dest=/tmp/lakefs.tar
+
+      - name: Upload artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: lakefs-image
+          path: /tmp/lakefs.tar
+
+  compatibility-checks-client:
+    name: Test lakeFS Hadoop FileSystem compatibility
+    needs: gen-code
+    strategy:
+      fail-fast: false
+      matrix:
+        # Removing a version from this list means the published client is no longer compatible with
+        # that lakeFS version.
+        lakeFS_version: [ 0.89.1, 0.90.0, 0.91.0, 0.92.0 ]
```

Comment:
so if there's a new version we need to remember to add it to the list?
can this be automated?

Reply:
Currently yes. It can be automated, I'm open to suggestions. 
On the other hand, I kind of feel this should be explicit "This version is compatible with master. This version isn't".

Reply:
it's just so prone to be forgotten... 
I wonder if, upon a new lakeFS release, this list can be updated automatically. 
We can save that list as a GitHub secret and append to it the most recent version when released. 
If any version becomes incompatible, it can be removed (by a workflow or by hand) from the list.

Reply:
A Github secret means that no one can see the current value of the secret and it's very hard to maintain. What value does the secret add on top of the list here? 

Version becoming incompatible is now a decision. How can a workflow decide if a version was broken due to a bug or known deprecation?

Reply:
It doesn't have to be a secret, it can be a GitHub variable and run as specified [here](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#example-using-contexts-to-create-matrices) (only for vars).
The value that I see being added is the automated part that is adding the **new** version to the list instead of us humans needing to remember or to follow a step in the release process spec which we'll, of course, forget at some point.
Version becoming incompatible will remain a decision: if an older version will need to be deleted this can be done manually or via another workflow.

Having said all that, it might be done as part of a new PR and not this one...